File: src/2024/day04/1/main.go
package main

import (
	_ "embed"
	"fmt"
	"strings"
)

//go:embed data.txt
var fileContents string

func findXsCount(line string) int {
	return strings.Count(line, "XMAS") + strings.Count(line, "SAMX")
}

func transposeMatrix(matrix [][]rune) [][]rune {
	rows, cols := len(matrix), len(matrix[0])
	transposed := make([][]rune, cols)
	for i := range transposed {
		transposed[i] = make([]rune, rows)
	}

	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			transposed[j][i] = matrix[i][j]
		}
	}
	return transposed
}

func getDiagonals(matrix [][]rune) [][]rune {
	rows, cols := len(matrix), len(matrix[0])
	diagonals := [][]rune{}

	// main and anti-diagonals
	for d := 1 - rows; d < cols; d++ {
		diag := []rune{}
		for r := 0; r < rows; r++ {
			c := d + r
			if c >= 0 && c < cols {
				diag = append(diag, matrix[r][c])
			}
		}
		if len(diag) > 0 {
			diagonals = append(diagonals, diag)
		}
	}

	// Reversed matrix diagonals
	reversedMatrix := make([][]rune, rows)
	for i := range matrix {
		reversedMatrix[i] = make([]rune, cols)
		for j := range matrix[i] {
			reversedMatrix[i][j] = matrix[i][cols-1-j]
		}
	}

	for d := 1 - rows; d < cols; d++ {
		diag := []rune{}
		for r := 0; r < rows; r++ {
			c := d + r
			if c >= 0 && c < cols {
				diag = append(diag, reversedMatrix[r][c])
			}
		}
		if len(diag) > 0 {
			diagonals = append(diagonals, diag)
		}
	}

	return diagonals
}

func main() {
	lines := strings.Split(strings.TrimSpace(fileContents), "\n")

	matrix := make([][]rune, len(lines))
	for i, line := range lines {
		matrix[i] = []rune(line)
	}

	var allLines [][]rune

	diagonals := getDiagonals(matrix)
	allLines = append(allLines, diagonals...)

	// rows
	allLines = append(allLines, matrix...)

	// columns
	columns := transposeMatrix(matrix)
	allLines = append(allLines, columns...)

	// count XS
	count := 0
	for _, lineRunes := range allLines {
		line := string(lineRunes)
		count += findXsCount(line)
	}

	fmt.Println(count)
}

File: src/2024/day04/1/main.rs
use regex::Regex;
use std::fs;
use std::path::Path;

fn read_lines() -> Vec<String> {
    let current_file = Path::new(file!());
    let current_dir = current_file
        .parent()
        .expect("Failed to get parent directory");
    let file_path = current_dir.join("data.txt");

    fs::read_to_string(file_path)
        .expect("Failed to read file")
        .lines()
        .map(|line| line.to_string())
        .collect()
}

fn find_xs_count(line: &str) -> usize {
    line.matches("XMAS").count() + line.matches("SAMX").count()
}

fn get_diagonals(matrix: &Vec<Vec<char>>) -> Vec<Vec<char>> {
    let rows = matrix.len();
    let cols = matrix[0].len();
    let mut diagonals = Vec::new();

    // main diagonals
    for d in (1 - rows as i32)..(cols as i32) {
        let mut diag = Vec::new();
        for r in 0..rows {
            let c = d + r as i32;
            if c >= 0 && c < cols as i32 {
                diag.push(matrix[r][c as usize]);
            }
        }
        if !diag.is_empty() {
            diagonals.push(diag);
        }
    }

    // reversed matrix diagonals
    let reversed_matrix: Vec<Vec<char>> = matrix
        .iter()
        .map(|row| row.iter().rev().cloned().collect())
        .collect();

    for d in (1 - rows as i32)..(cols as i32) {
        let mut diag = Vec::new();
        for r in 0..rows {
            let c = d + r as i32;
            if c >= 0 && c < cols as i32 {
                diag.push(reversed_matrix[r][c as usize]);
            }
        }
        if !diag.is_empty() {
            diagonals.push(diag);
        }
    }

    diagonals
}

fn transpose(matrix: &Vec<Vec<char>>) -> Vec<Vec<char>> {
    let rows = matrix.len();
    let cols = matrix[0].len();

    (0..cols)
        .map(|c| (0..rows).map(|r| matrix[r][c]).collect())
        .collect()
}

fn main() {
    let lines = read_lines();
    let matrix: Vec<Vec<char>> = lines.iter().map(|line| line.chars().collect()).collect();

    let mut all_lines = Vec::new();

    // diagonals
    all_lines.extend(get_diagonals(&matrix));

    // rows
    all_lines.extend(matrix.clone());

    // columns
    all_lines.extend(transpose(&matrix));

    let count: usize = all_lines
        .iter()
        .map(|line| find_xs_count(&line.iter().collect::<String>()))
        .sum();

    println!("{}", count);
}
